### PR TITLE
feat: add extemporaneous grading feature

### PR DESCRIPTION
**Pull request by BryanttV**
### Description
This PR adds the extemporaneous grading feature. Instructors can define a **due date** and **late due date** in the component. Learners can access the children's content of the main component or not depending on the dates predefined by the instructor. The following fields were included in the component configuration:

* **Due Date**: Allows to configure the due date of the component.
* **Due Time (UTC)**: Allows to configure the due time of the component.
* **Late Due Date**: Allows to configure the late due date of the component.
* **Late Due Time (UTC)**: Allows to configure the late due time of the component.
* **Due Date Explanation Text**: Allows to configure the text displayed to learners when the due date has passed.
* **Late Due Date Explanation Text**: Allows to configure the text displayed to learners when the late due date has passed.

### How to Test
1. Install this branch.
2. Add `extemporaneous_grading` to **Studio** > **Course** > **Settings** > **Advanced Settings**.
3. Configure a new component of **Extemporaneous Grading** in a unit.
4. From LMS:
    - If the **due date** has not passed, you will be able to access the content of the component.
    - If the **due date** has passed, you will be able to access the content of the component if you accept that the submission is late.
    - If the **late due date** has passed, you will not be able to access the content of the component.

### CODE REVIEW

**Comment by BryanttV** [comment]
I will be working on correcting the unit tests.

File: extemporaneous_grading/conf/locale/es_419/LC_MESSAGES/text.po [hunk 1]
```diff
@@ -17,14 +17,122 @@ msgstr ""
...
+"entrega tardía haya pasado."
+
+#: extemporaneous_grading.py:98
+msgid ""
+"The late due date has passed. You can not submit this assignment anymore."
+msgstr ""
+"La fecha de entrega tardía ha pasado. Ya no puedes entregar esta tarea."
+
+#: extemporaneous_grading.py:102
+msgid "Is Late Submission"
```
Comment (mariajgrimaldi) [inline extemporaneous_grading/conf/locale/es_419/LC_MESSAGES/text.po]: The grammatical correct way is:
Is it a late submission?
So I think we can use: Late Submission instead

File: extemporaneous_grading/extemporaneous_grading.py [hunk 2]
```diff
@@ -38,8 +44,75 @@ class XBlockExtemporaneousGrading(StudioContainerWithNestedXBlocksMixin, StudioE
...
+        default="00:00",
+    )
+
+    late_due_date = DateTime(
+        display_name=_("Late Due Date"),
+        help=_(
+            "The late due date for this component. If the learner access after the due date "
+            "and before the late due date, the learner will see a message that the submission "
+            "is late. If the learner accepts the late submission pressing the button, the "
+            "learner will see the content."
```
Comment (mariajgrimaldi) [inline extemporaneous_grading/extemporaneous_grading.py]:
```suggestion
            "The late due date for this component. If the learner accesses it after the due date "
            "and before the late due date, the learner will see a message that the submission "
            "is late. If the learner accepts the late submission by pressing the button, the "
            "learner will see the content."
```

File: extemporaneous_grading/extemporaneous_grading.py [hunk 3]
```diff
@@ -38,8 +44,75 @@ class XBlockExtemporaneousGrading(StudioContainerWithNestedXBlocksMixin, StudioE
         default=_("Extemporaneous Grading"),
     )
 
+    due_date = DateTime(
+        display_name=_("Due Date"),
+        help=_(
+            "The due date for this component. If the learner access before this date, "
+            "the learner will see the content without any restrictions."
```
Comment (mariajgrimaldi) [inline extemporaneous_grading/extemporaneous_grading.py]:
```suggestion
            "The due date for this component. If the learner accesses it before this date, "
            "the learner will see the content without any restrictions."
```

File: extemporaneous_grading/extemporaneous_grading.py [hunk 4]
```diff
@@ -38,8 +44,75 @@ class XBlockExtemporaneousGrading(StudioContainerWithNestedXBlocksMixin, StudioE
...
+    late_due_time = String(
+        display_name=_("Late Due Time (UTC)"),
+        help=_("The time of the late due date. The valid format is HH:MM."),
+        scope=Scope.settings,
+        default="00:00",
+    )
+
+    due_date_explanation_text = String(
+        display_name=_("Due Date Explanation Text"),
+        help=_("The explanation text that will be shown to the learner when the due date is passed."),
```
Comment (mariajgrimaldi) [inline extemporaneous_grading/extemporaneous_grading.py]:
```suggestion
        help=_("The explanation text that will be shown to the learner when the due date has passed."),
```

File: extemporaneous_grading/extemporaneous_grading.py [hunk 5]
```diff
@@ -38,8 +44,75 @@ class XBlockExtemporaneousGrading(StudioContainerWithNestedXBlocksMixin, StudioE
...
+        default=_(
+            "The due date has passed. You can still submit this assignment until the "
+            "late due date. After the late due date, you will not be able to submit "
+            "this assignment. If you accept the late submission, press the button."
+        ),
+    )
+
+    late_due_date_explanation_text = String(
+        display_name=_("Late Due Date Explanation Text"),
+        help=_("The explanation text that will be shown to the learner when the late due date is passed."),
```
Comment (mariajgrimaldi) [inline extemporaneous_grading/extemporaneous_grading.py:98]:
```suggestion
        help=_("The explanation text that will be shown to the learner when the late due date has passed."),
```

File: extemporaneous_grading/static/css/extemporaneous_grading.css [hunk 6]
```diff
@@ -1 +1,11 @@
...
+
+.dates {
+    display: grid;
+    margin: 0 0 15px 0;
+    color: rgb(94, 94, 94);
+}
+
+.dates span {
+    margin: 5px 0;
+}
```
Comment (mariajgrimaldi) [inline extemporaneous_grading/static/css/extemporaneous_grading.css]: Missing EOL

File: extemporaneous_grading/static/html/children.html [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    </div>
+    <hr />
+    {% for child_content in children_contents %}
+        {{ child_content|safe }}
+        <br />
+        <hr />
+    {% endfor %}
+</div>
+
+<script type="text/javascript" src="/extemporaneous_grading/static/js/src/resize_iframe.js"></script>
```
Comment (mariajgrimaldi) [inline extemporaneous_grading/static/html/children.html]: I thought this was being added here regardless: https://github.com/eduNEXT/xblock-extemporaneous-grading/pull/2/files#diff-e60628835ced9b5ca0bccf5fd0ed4e046442e626c2fc5cd9e8127555307fc1d7R244
Should this be done manually too?

File: extemporaneous_grading/static/html/late_due_datetime.html [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{% load i18n %}
+<div class="extemporaneous_grading_block">
+    <p>{% trans block.late_due_date_explanation_text %}</p>
+</div>
+
+<script type="text/javascript" src="/extemporaneous_grading/static/js/src/resize_iframe.js"></script>
```
Comment (mariajgrimaldi) [inline extemporaneous_grading/static/html/late_due_datetime.html]: ^ Same question here

**Review by mariajgrimaldi — CHANGES_REQUESTED** [review]

File: extemporaneous_grading/conf/locale/es_419/LC_MESSAGES/text.po (hunk 1 above)
Comment (BryanttV) [reply extemporaneous_grading/conf/locale/es_419/LC_MESSAGES/text.po]: Done: https://github.com/eduNEXT/xblock-extemporaneous-grading/commit/aaed9daa91de0eebf250d0cba14446801701ab82

File: extemporaneous_grading/extemporaneous_grading.py (hunk 2 above)
Comment (BryanttV) [reply extemporaneous_grading/extemporaneous_grading.py]: Done: https://github.com/eduNEXT/xblock-extemporaneous-grading/commit/a23b2c0fea2db1db0c613abdd3814bdb892c311a

File: extemporaneous_grading/extemporaneous_grading.py (hunk 3 above)
Comment (BryanttV) [reply extemporaneous_grading/extemporaneous_grading.py]: Done: https://github.com/eduNEXT/xblock-extemporaneous-grading/commit/a23b2c0fea2db1db0c613abdd3814bdb892c311a

File: extemporaneous_grading/extemporaneous_grading.py (hunk 4 above)
Comment (BryanttV) [reply extemporaneous_grading/extemporaneous_grading.py]: Done: https://github.com/eduNEXT/xblock-extemporaneous-grading/commit/a23b2c0fea2db1db0c613abdd3814bdb892c311a

File: extemporaneous_grading/extemporaneous_grading.py (hunk 5 above)
Comment (BryanttV) [reply extemporaneous_grading/extemporaneous_grading.py:98]: Done: https://github.com/eduNEXT/xblock-extemporaneous-grading/commit/a23b2c0fea2db1db0c613abdd3814bdb892c311a

File: extemporaneous_grading/static/css/extemporaneous_grading.css (hunk 6 above)
Comment (BryanttV) [reply extemporaneous_grading/static/css/extemporaneous_grading.css]: Done: https://github.com/eduNEXT/xblock-extemporaneous-grading/commit/736dc12d0e9af661fe43f870a8da58c2d450702e

File: extemporaneous_grading/static/html/children.html (hunk 7 above)
Comment (BryanttV) [reply extemporaneous_grading/static/html/children.html]: That's right! We don't need to add it manually. Thank you!

Done: https://github.com/eduNEXT/xblock-extemporaneous-grading/commit/0b0a990f9fb3e1630c7dd8e1673fe1eca47f1002

File: extemporaneous_grading/static/html/late_due_datetime.html (hunk 8 above)
Comment (BryanttV) [reply extemporaneous_grading/static/html/late_due_datetime.html]: Done: https://github.com/eduNEXT/xblock-extemporaneous-grading/commit/0b0a990f9fb3e1630c7dd8e1673fe1eca47f1002

File: extemporaneous_grading/extemporaneous_grading.py [hunk 9]
```diff
@@ -38,8 +44,77 @@ class XBlockExtemporaneousGrading(StudioContainerWithNestedXBlocksMixin, StudioE
...
+        display_name=_("Late Due Date"),
+        help=_(
+            "The late due date for this component. If the learner access it after the due date "
+            "and before the late due date, the learner will see a message that the submission "
+            "is late. If the learner accepts the late submission by pressing the button, the "
+            "learner will see the content. When the late due date passes, the learner will not "
+            "be able to see the content."
+        ),
+        scope=Scope.settings,
+        default=datetime.now(),
```
Comment (mariajgrimaldi) [inline extemporaneous_grading/extemporaneous_grading.py:74]: why do we need a default here?

**Review by mariajgrimaldi — APPROVED** [review]

**Pull request merged**